### PR TITLE
Multiple Sidebar Panel Layouts

### LINKED ISSUE #26706: Option to hide nearby monsters indicator.
A picture is better than a thousand words (https://imgur.com/a/ZYm0mZ9) . As you can see, with my current font settings and monitor (1366x768 laptop screen, font size 20) I get around 137x37 terminal size. When the minimap is active, there is too little room left for the message log (about 2-4 lines), which is way to little, and I can miss important messages.

I'm not sure how that area is called, so I call it monsters indicator. It is very useful, but some of it's job is overtaken by the mini-map. I'd like to request an option or a key to hide it. Alternatively, it could toggle between 3 states: normal, hidden and "single line" where you only display ONLY the symbols of nearby off-screen monsters. If I see a purple "&" there, I sure will know that there's a mi-go nearby! And I can always press "V" to check.

### LINKED ISSUE #27367: sidebar needs an overhaul
**Issue:**  
Sidebar take too much space. 
The overall feeling is messy it needs more order.
The sidebar code also need cleaning.

**Proposal**  
The idea is to get independent panels, that can get toggled on and off by the user. 
those panels could be stacked to simulate the old sidebar. 
and in the end the user could choose which panels he want to display. 
for example the user could do: 
- 0 panel on screen (just opening them on a need basis then immediately close them)
- 1 panel only (keeping most of the time only one panel open)
- 2, 3, 4.. panels on screen. 
- simulate the old sidebar by stacking a few panels of his choice and keep them most of the time.
- those panels could of course be toggled on and off at will while playing.
- choose the spacing between panels

Panels "a la carte" ..
Please consider the following mockups. 
( the term dimension is 120x37, font is 20x20 that makes a resolution of 1440x900 )
preview2
preview11
preview8

the following mockups represent a minimal dimension (80x25) terminal. 
( my font size is 20x20 so that make resolution of 1600x500 )
most users will not use such minimal setting of 80x25 terminal dimension. 
it is shown to preview what this idea would look like for those playing on lower dimensions terms.
it allow to maximize game area, by showing only 
what we need when we need and not 'everything' all the time.
preview3
preview9
preview10

**Coding**
Iam already trying to do it. This is a follow up from my [pull request](https://github.com/CleverRaven/Cataclysm-DDA/pull/27352) on cleaning the existing sidebar. While doing it, i came up with the idea and did the mockup. The problem is i'm not a very skilled cpp dev just yet. So it will take me time. If someone want to help or takeover the idea then great, do it. 
Otherwise i'll continue to do it myself slowly at my pace. I just wanted to put the idea out and eventually get a bit of help and feedback.
For the beginning i'm planning on making 5 or 6 panel available. 
The user will choose when and what to display, as shown in the mockup.
The good new is that most code already exists for all this.

planned panels:
- character
- environment
- messages
- compass
- minimap
- look_around / mouse_view

Since everything will be segmented, each panel could have its own function.
It mean everything will get cleaner. And we can also load them all from a central point, 
if conditions are met (if user toggle them on).
```
load_panels() 
{
    if (toggle_compass)
        draw_compass(); 
    if (toggle_minimap)
        draw_minimap(); 
    if (toggle_messages)
        draw_messages();
    if (toggle_environment)
        draw_environment();
    if (toggle_character)
        draw_character();
}
```

### PR DESCRIPTION
**Pull request by KorGgenT**
#### Summary

```SUMMARY: Interface "Multiple Sidebar Panel Layouts"```
#### Purpose of change

Closes #26706

Closes #27367
Makes the sidebar much more customizable.
#### Describe the solution

Uses @nsklaus 's idea of the rearrange-able panel system in the sidebar, and expands that to have the ability to have multiple layouts.
#### Additional context

"Classic":
image

image

image

"Compact":
image

image

image

"Labels": (panel in screenshot is the window that allows panel selection)
image

image

image

Currently does not save settings when you close the game.
Will want some input from translators.
- [x] compass?
- [x] Center panel admin.
- [x] Pixel minimap

### CODE REVIEW

**Comment by OrenAudeles** [comment]
Does this work in Curses builds or is it Tiles only?

**Comment by KorGgenT** [comment]
still need to try curses, this definitely works in tiles

**Comment by Amneiger** [comment]
I said this on Reddit already, but I'm very happy to see a player-customizable UI.

My C++ is too substandard to figure this out: which option is the default that's used when loading the game?

The screenshot for the classic vehicle sidebar is the same as the screenshot for the compact vehicle sidebar.  For everyone's peace of mind, could you take a moment to get that fixed?

**Comment by KorGgenT** [comment]
> I said this on Reddit already, but I'm very happy to see a player-customizable UI.
> 
> My C++ is too substandard to figure this out: which option is the default that's used when loading the game?
> 
> The screenshot for the classic vehicle sidebar is the same as the screenshot for the compact vehicle sidebar. For everyone's peace of mind, could you take a moment to get that fixed?

"classic" is default when loading the game

**Comment by Amneiger** [comment]
> "classic" is default when loading the game

Got it, thanks.

It looks like you changed the screenshots?  The current classic vehicle screenshot doesn't match what I see in a pre-UI changes build:

classic vehicle sidebar

I use Safe mode in vehicles to help me notice nearby mi-gos and crash sites when driving in new places, or to see when I'm getting too close to an uncleared town, so the absence of it stood out to me.  I also noticed the lack of temperature, because I like to know when it's okay to plant seeds or immediately notice if I've stepped into an ice lab, and then I just don't bother putting the thermometer down.

**Comment by KorGgenT** [comment]
> > "classic" is default when loading the game
> 
> Got it, thanks.
> 
> It looks like you changed the screenshots? The current classic vehicle screenshot doesn't match what I see in a pre-UI changes build:
> 
> classic vehicle sidebar
> 
> I use Safe mode in vehicles to help me notice nearby mi-gos and crash sites when driving in new places, or to see when I'm getting too close to an uncleared town, so the absence of it stood out to me. I also noticed the lack of temperature, because I like to know when it's okay to plant seeds or immediately notice if I've stepped into an ice lab, and then I just don't bother putting the thermometer down.

I see. That's a pretty easy fix. I sort of moved the fuel gauge because i thought it would be better to have more information (i.e. your stats) while driving, and having safe mode on is pretty much irrelevant, though i guess i can put it back to be totally the old style. also the temp does show, i just only had a thermometer on me in the third screenshot (in the lab)

I will also say that once the ability to save the settings is put in (no promises it will be this PR) the default will be made into "labels" since labels make it a bit easier for new players.

**Comment by Amneiger** [comment]
>i guess i can put it back to be totally the old style

Folks are a bit twitchy about this at the moment.  I'm going to suggest that classic be as close to the original UI as much as possible.  

>I will also say that once the ability to save the settings is put in (no promises it will be this PR) the default will be made into "labels" since labels make it a bit easier for new players.

That's fair, since people will be able to save their settings and not have to use the other UIs if they don't want to.

**Comment by KorGgenT** [comment]
I have made edits until I can't see any more edits to make. For those of you following this PR and do not know how to compile, and also run windows, I have made a [distributable](https://drive.google.com/open?id=1uXlKPW3tG24aUEcQok-KaW884_g2pe5x) for you to check out.

**Comment by AMurkin** [comment]
`Manage zones` and `List all items around` have display issues.

**Comment by AMurkin** [comment]
Also `Sidebar position` (Right/Left) option does nothing now. And `game.h` have unused `bool right_sidebar;`

**Comment by KorGgenT** [comment]
@AMurkin I fixed the sidebar position thing, would you elaborate on the display issues you are referring to?

**Comment by AMurkin** [comment]
### Manage zones
No zones around:

Now | PR
---|---
default | default

If there is a zone defined, inner part of the zones manager is momentarily displayed and erased.

### List all items around

Now | PR
---|---
default | default

**Comment by AMurkin** [comment]
When I scroll items in List all items around window sometimes it didn't draw target marker to item's position. Also if sidebar position is set to Left, it hides sometimes.

Zones doesn't highlighted now. In the master version their tiles are blinking.
No border here at the bottom-right:
default

**Comment by I-am-Erk** [comment]
Does this solve https://github.com/CleverRaven/Cataclysm-DDA/pull/28152 ?

**Comment by KorGgenT** [comment]
not by itself

File: src/game.cpp [hunk 1]
```diff
@@ -7812,27 +7451,23 @@ void centerlistview( const tripoint &active_item_position )
...
-                u.view_offset.x = 0;
-                u.view_offset.y = 0;
+                u.view_offset.y = ypos - ( TERRAIN_WINDOW_HEIGHT - 1 ) + yOffset;
             }
         } else {
             if( xpos < 0 ) {
                 u.view_offset.x = xpos;
-            } else if( xpos >= TERRAIN_WINDOW_WIDTH ) {
-                u.view_offset.x = xpos - ( TERRAIN_WINDOW_WIDTH - 1 );
+            // magic number is because terrain window is drawn underneath the sidebar
```
Comment (AMurkin) [inline src/game.cpp]:
```suggestion
                // magic number is because terrain window is drawn underneath the sidebar
```

File: src/panels.cpp [hunk 2]
```diff
@@ -0,0 +1,2013 @@
...
+    // weather
+    if( g->get_levz() < 0 ) {
+        mvwprintz( w, 3, 8, c_light_gray, _( "Underground" ) );
+    } else {
+        mvwprintz( w, 3, 8, weather_data( g->weather ).color, weather_data( g->weather ).name.c_str() );
+    }
+    // display lighting
+    const auto ll = get_light_level( g->u.fine_detail_vision_mod() );
+    mvwprintz( w, 4, 8, ll.second, ll.first.c_str() );
+    // wind
```
Comment (davidpwbrown) [inline src/panels.cpp]:
```suggestion
    // wind
    const oter_id &cur_om_ter = overmap_buffer.ter( p->global_omt_location() );
    double windpower = get_local_windpower( g->windspeed, cur_om_ter, 
               p->pos(), g->winddirection, g->is_sheltered( p->pos() ) );
    mvwprintz( w, 5, 8, get_wind_color( g->windspeed ),
               get_wind_desc( windpower ) + " " + get_wind_arrow( g->winddirection ) );

```

**Comment by KorGgenT** [comment]
all right! i've gotten out all the kinks now! i don't need to make any more edits unless someone finds a bug.

**Comment by KorGgenT** [comment]
jenkins rebuild

**Comment by kevingranade** [comment]
It's quite flickery, and it leaves ghosts a lot when adjusting the sidebar, but I think we can fix that up later considering how huge this has to be.  I need to spend some time reading the code, but everything functional looks good.

File: src/mission_companion.cpp [hunk 3]
```diff
@@ -616,6 +619,31 @@ npc_ptr talk_function::individual_mission( const tripoint &omt_pos,
     return comp;
 }
 
+std::vector<item *> talk_function::individual_mission_give_equipment( std::vector<item *> equipment,
+        const std::string &message )
+{
+    std::vector<item *> equipment_lost;
```
Comment (kevingranade) [inline src/mission_companion.cpp]: Merge error? This seems unrelated to the sidepanel.

Comment (KorGgenT) [reply src/mission_companion.cpp]: hm yes, it's probably a mistake from rebasing it in order to remove conflicts.

**Comment by rumly111** [comment]
I get segfault when I try to turn on the pixel minimap

**Comment by dommon** [comment]
> still need to try curses, this definitely works in tiles

yea, get back with that before clearing to merge, oh wait..

**Comment by KorGgenT** [comment]
> > still need to try curses, this definitely works in tiles
> 
> yea, get back with that before clearing to merge, oh wait..

sorry. we spoke about it on discord at the time. it worked.

File: src/game.cpp [hunk 4]
```diff
@@ -511,159 +503,32 @@ void game::init_ui( const bool resized )
...
-        // The default style only uses one status window.
-        stat2X = 0;
-        stat2Y = statY + statH;
-        stat2H = 1;
-        stat2W = sidebarWidth;
+    minimapX = 0;
+    minimapY = 0;
+    pixelminimapW = MINIMAP_WIDTH;
+    pixelminimapH = ( pixelminimapW / 2 );
+    if( pixel_minimap_custom_height &&
```
Comment (kevingranade) [inline src/game.cpp]: Maybe here

File: src/panels.cpp [hunk 5]
```diff
@@ -0,0 +1,2015 @@
...
+        return c_white;
+    } else {
+        return c_green;
+    }
+}
+
+int window_panel::get_height() const
+{
+    if( height == -1 ) {
+        return get_option<int>( "PIXEL_MINIMAP_HEIGHT" );
```
Comment (kevingranade) [inline src/panels.cpp]: Maybe here

**Comment by davidpwbrown** [comment]
This will cause problems with editmap()
I've fixed one crash from faction camps callign editmap.
There will likely be others ( int in discord has found one )
Probably with debug map editing and anything that uses editmap.
`width = TERMX - TERRAIN_WINDOW_TERM_WIDTH;` in editmap() will be zero now, it seems, when before it wasnt.
and a zero width will cause problems.

**Comment by paulenka-aleh** [comment]
@KorGgenT 

How do I enter sidebar options menu?

Sorry for a dumb question, just downloaded latest experimental, copied previous settings and after playing for 5 minutes realized minimap is too way small for me and some menus float to the left (e.g. pickup) while I used to focus on the right side while paying the game. I tried googling, checking help, checking keybindings, going through options but didn't manage to find a way. I didn't dived into the code yet though...
